azurelinuxagent/ga/ga_version_updater.py

```diff
@@ -104,6 +98,33 @@ def sync_new_gs_id(self, gs_id):
         """
         self._gs_id = gs_id
 
+    @staticmethod
+    def download_new_agent_pkg(package_to_download, protocol, is_fast_track_goal_state):
+        """
+        Function downloads the new agent.
+        @param package_to_download: package to download
+        @param protocol: protocol object
+        @param is_fast_track_goal_state: True if goal state is fast track else False
+        """
+        agent_name = "{0}-{1}".format(AGENT_NAME, package_to_download.version)
+        agent_dir = os.path.join(conf.get_lib_dir(), agent_name)
+        agent_pkg_path = ".".join((os.path.join(conf.get_lib_dir(), agent_name), "zip"))
+        agent_handler_manifest_file = os.path.join(agent_dir, AGENT_MANIFEST_FILE)
+        if not os.path.exists(agent_dir) or not os.path.isfile(agent_handler_manifest_file):
+            protocol.client.download_zip_package("agent package", package_to_download.uris, agent_pkg_path, agent_dir, use_verify_header=is_fast_track_goal_state)
+        else:
+            logger.info("Agent {0} was previously downloaded - skipping download", agent_name)
+
+        if not os.path.isfile(agent_handler_manifest_file):
```

Comment:
I still find this check confusing: why do we check for the manifest, but nor for the egg? we need both in order to execute the Agent. Or looking at this differently, do we need to check at all? the download/zip already succeed at this point.

Reply:
Intention is not to exit the current process when we know we can't run the agent without both. I agree we need both checks or could continue with update since download/zip succeeded. When Deamon finds missing, it will eventually blacklist and stop the update.  Former case, we may end up downloading agent again and again if it's more permanent issue. Latter case, we don't give second chance to download. Not see an advantage in either case though.:)

Reply:
well, latter case not having check safer than download again?

Reply:
I am not sure I follow the "downloading agent again and again" vs "we don't give second chance to download.". 

This may be related to my other comment about checking that the directory exists and skipping the download. Every update attempt should start from scratch, I think, to prevent these issues of incomplete downloads.

If we really want to check the integrity of the downloaded agent, the check should be stronger than checking for the manifest (or the egg, or both), more like adding a checksum to the package and verifying the checksum.